app/src/main/java/org/groundplatform/android/ui/datacollection/DataCollectionViewModel.kt

```diff
@@ -99,72 +115,94 @@ internal constructor(
+  private val taskViewModels: MutableStateFlow<MutableMap<String, AbstractTaskViewModel>> =
+    MutableStateFlow(mutableMapOf())
+
+  private val taskDataHandler = TaskDataHandler()
+  private lateinit var taskSequenceHandler: TaskSequenceHandler
 
-  val surveyId: String = requireNotNull(surveyRepository.activeSurvey?.id)
+  private val currentTaskId: StateFlow<String> = savedStateHandle.getStateFlow(TASK_POSITION_ID, "")
 
   val jobName: StateFlow<String> =
-    MutableStateFlow(job.name ?: "").stateIn(viewModelScope, SharingStarted.Lazily, "")
-
-  val loiName: StateFlow<String?> =
-    (if (loiId == null) {
-        // User supplied LOI name during LOI creation task. Use to save the LOI name later.
-        savedStateHandle.getStateFlow(TASK_LOI_NAME_KEY, "")
-      } else
-      // LOI name pulled from LOI properties, if it exists.
-      flow {
-          locationOfInterestRepository.getOfflineLoi(surveyId, loiId)?.let {
-            val label = locationOfInterestHelper.getDisplayLoiName(it)
-            emit(label)
+    loadState
+      .map { (it as? LoadState.Ready)?.job?.name.orEmpty() }
+      .distinctUntilChanged()
+      .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), "")
+
+  val loiName: StateFlow<String> =
+    if (loiId == null) {
+      savedStateHandle.getStateFlow(TASK_LOI_NAME_KEY, "")
+    } else {
+      surveyRepository.activeSurveyFlow
+        .filterNotNull()
+        .map { survey ->
+          withContext(ioDispatcher) {
+            locationOfInterestRepository
+              .getOfflineLoi(survey.id, loiId)
+              ?.let { locationOfInterestHelper.getDisplayLoiName(it) }
+              .orEmpty()
           }
-        })
-      .stateIn(viewModelScope, SharingStarted.Lazily, "")
+        }
+        .distinctUntilChanged()
+        .stateIn(viewModelScope, SharingStarted.Lazily, "")
+    }
 
   val loiNameDialogOpen: MutableState<Boolean> = mutableStateOf(false)
 
-  private val taskViewModels: MutableStateFlow<MutableMap<String, AbstractTaskViewModel>> =
-    MutableStateFlow(mutableMapOf())
+  init {
+    viewModelScope.launch {
+      _loadState.value = LoadState.Loading
+      val result = runCatching {
+        withTimeout(SURVEY_LOAD_TIMEOUT_MILLIS) {
+          val survey = surveyRepository.activeSurveyFlow.filterNotNull().first()
+          val job = survey.getJob(jobId) ?: error("couldn't retrieve job for $jobId")
+          val tasks =
+            if (isAddLoiFlow) job.tasksSorted else job.tasksSorted.filterNot { it.isAddLoiTask }
+          LoadState.Ready(survey = survey, job = job, tasks = tasks, surveyId = survey.id)
+        }
+      }
 
-  private val taskDataHandler = TaskDataHandler()
-  private val taskSequenceHandler = TaskSequenceHandler(tasks, taskDataHandler)
+      result
+        .onSuccess { ready ->
+          this@DataCollectionViewModel.job = ready.job
+          this@DataCollectionViewModel.tasks = ready.tasks
```

Comment:
Can we remove the variables `job` and `tasks` and directly get the value from the `loadState`? This would ensure that we won't have to worry about resetting the variables if the state changes. Wdyt?